openfl/callbacks/callback.py

```diff
@@ -0,0 +1,56 @@
+# Copyright 2020-2024 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+
+class Callback:
+    """Base class for callbacks.
+
+    Callbacks can be used to perform actions at different stages of the
+    Federated Learning process. To create a custom callback, subclass
+    `openfl.callbacks.Callback` and implement the necessary methods.
+
+    Callbacks can be triggered on the aggregator and collaborator side
+    for the following events:
+        * At the beginning of an experiment
+        * At the beginning of a round
+        * At the end of a round
+        * At the end of an experiment
+
+    Attributes:
+        params: Additional parameters saved for use within the callback.
+        tensor_db: The `TensorDB` instance of the respective participant.
+    """
+
+    def __init__(self):
+        self.params = None
+        self.tensor_db = None
+
+    def set_params(self, params):
+        self.params = params
+
+    def set_tensor_db(self, tensor_db):
+        self.tensor_db = tensor_db
+
+    def on_round_begin(self, round_num: int, logs=None):
```

Comment:
Based on my read through, I see that the `logs` named-arg is only used by `MetricWriter`. 
For extensibility, would it make more sense to have a `**kwargs` instead?
We would still need to consider how these "dynamic" named args are sent by the aggregator/collaborator but this could be maybe thought of in subsequent/future changes to the `callbacks` API.

Reply:
> For extensibility, would it make more sense to have a **kwargs instead?

Generally speaking, `**kwargs` should be avoided. To pass information that specific callbacks would need, we have `params`  that are declared at the time of instantiating `callbacks` list. This makes it clear what attributes `callbacks` will have access to, internally. Dynamic setting makes it harder to inspect and test code.

https://github.com/securefederatedai/openfl/blob/38f7c30c57aa5d92db1ecf0c35a22be6733ed30d/openfl/callbacks/callback_list.py#L18-L20

Currently, `logs` are the default because most `callbacks` generally trigger actions based on metrics - say writing `logs` to file, trigger saving a checkpoint if `logs["loss"] < 0.2`, write `logs` to tensorboard etc. 

That said, I agree that we can expand this argument list further - if more callbacks benefit from them.